client.go

```diff
@@ -1,82 +1,235 @@
+		turnServ, err = config.Net.ResolveUDPAddr("udp4", config.TURNServerAddr)
+		if err != nil {
+			return nil, err
+		}
+		log.Debugf("turnServ: %s", stunServ.String())
 	}
 
-	var err error
-	c.conn, err = c.net.ListenPacket(network, config.ListeningAddress)
-	if err != nil {
-		return nil, errors.Wrap(err, fmt.Sprintf("failed to listen on %s", config.ListeningAddress))
+	c := &Client{
+		conn:     config.Conn,
+		stunServ: stunServ,
+		turnServ: turnServ,
+		username: stun.NewUsername(config.Username),
+		password: config.Password,
+		realm:    stun.NewRealm(config.Realm),
+		net:      config.Net,
+		trMap:    map[string]*transaction{},
+		rto:      defaultRTO,
+		log:      log,
 	}
 
 	return c, nil
 }
 
-// SendSTUNRequest sends a new STUN request to the serverIP with serverPort
-func (c *Client) SendSTUNRequest(serverIP net.IP, serverPort int) (net.Addr, error) {
-	c.mux.Lock()
-	defer c.mux.Unlock()
+// TURNServerAddr return the TURN server address
+func (c *Client) TURNServerAddr() net.Addr {
+	return c.turnServ
+}
 
-	c.log.Debug("sending STUN request")
-	if err := sendStunRequest(c.conn, serverIP, serverPort); err != nil {
-		return nil, err
+// STUNServerAddr return the STUN server address
+func (c *Client) STUNServerAddr() net.Addr {
+	return c.stunServ
+}
+
+// Username returns username
+func (c *Client) Username() stun.Username {
+	return c.username
+}
+
+// Realm return realm
+func (c *Client) Realm() stun.Realm {
+	return c.realm
+}
+
+// WriteTo sends data to the specified destination using the base socket.
+func (c *Client) WriteTo(data []byte, to net.Addr) (int, error) {
+	return c.conn.WriteTo(data, to)
+}
+
+// Listen will have this client start listening on the conn provided via the config.
+// This is optional. If not used, you will need to call HandleInbound method
+// to supply incoming data, instead.
+func (c *Client) Listen() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.conn == nil {
+		conn, err := c.net.ListenPacket("udp4", "0.0.0.0:0")
+		if err != nil {
+			return err
+		}
+		c.conn = conn
 	}
 
-	packet := make([]byte, 1500)
-	c.log.Debug("wait for STUN response...")
-	size, _, err := c.conn.ReadFrom(packet)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to read packet from udp socket")
+	if c.isListening {
+		return fmt.Errorf("already listening")
+	}
+	c.isListening = true
+
+	go func() {
+		buf := make([]byte, maxDataBufferSize)
+		for {
+			n, from, err := c.conn.ReadFrom(buf)
+			if err != nil {
+				c.log.Debugf("exiting read loop: %s", err.Error())
+				break
+			}
+
+			_, err = c.HandleInbound(buf[:n], from)
+			if err != nil {
+				c.log.Debugf("exiting read loop: %s", err.Error())
+				break
+			}
+		}
+
+		c.mutex.Lock()
+		c.isListening = false
+		c.mutex.Unlock()
+	}()
+
+	return nil
+}
+
+// Close closes this client
+func (c *Client) Close() {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.isListening {
+		c.conn.Close() // nolint:errorcheck
```

Comment:
Can we return this to the user, we can just put at the end of the function
```
if c.isListening {
  return c.conn.Close()
}

return nil
```

Reply:
conn is no longer closed by the client.